services/util/array-utils/array.utils.go

```diff
@@ -0,0 +1,27 @@
+/* Copyright (c) 2024 Snowflake Inc. All rights reserved.
+
+   Licensed under the Apache License, Version 2.0 (the
+   "License"); you may not use this file except in compliance
+   with the License.  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing,
+   software distributed under the License is distributed on an
+   "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+   KIND, either express or implied.  See the License for the
+   specific language governing permissions and limitations
+   under the License.
+*/
+
+package array_utils
+
+// FindIndexBy finds the index of value in array
+func FindIndexBy[T comparable](arr []T, predicate func(val T) bool) int {
```

Comment:
continuing on https://github.com/Snowflake-Labs/sansshell/pull/489#discussion_r1781441781

my preference is to avoid creating new Go packages until there's an actual use case for it.

Having more packages would make the codebase harder to follow.

Also, https://dave.cheney.net/2019/01/08/avoid-package-names-like-base-util-or-common

Reply:
> Having more packages would make the codebase harder to follow.

But have the code in place make it harder to understand. It is much simpler to undestand what is going on when you see
```golang
index := arrUtils.FindIndexBy(arr, func (val) => val == 5)
// vs
index := -1
for i, val := range arr {
  if val == 5 {
     index := i
  }
}
```
`FindIndexBy` is super generic and I definitely sure that I reuse it

Reply:
I see several problems with this approach, especially when we scale it to each situation:
- when it would require to reuse it in different module it would require to move function and tests what would increase code changes in PR
- for me it would require from me to remember all places where I make some handy utils to being able to reuse them
- for other devs, highly likely, they won't even know that such utility existing and implement their own. In this case there 2 sub scenario:
   - I need to reject on code review and force them to refactor some code, what increase task scope and delivery time
   - we would have a lot similar function across project, but different enough to make refactoring. Together with other factors it increase chance of regression

On the other hand I don't see solid benefits to not doing this